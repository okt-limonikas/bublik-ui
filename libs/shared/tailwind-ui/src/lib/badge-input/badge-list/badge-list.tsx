/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { useRef, RefObject } from 'react';

import { BadgeItem } from '../types';
import { EditPopover } from './edit-popover';
import { cn } from '../../utils';

export interface BadgeListProps {
	label?: string;
	badges: BadgeItem[];
}

export const BadgeList = ({ label, badges }: BadgeListProps) => {
	const listRef = useRef<HTMLUListElement>(null);

	return (
		<ul
			className={cn(
				'flex flex-wrap items-center gap-1',
				badges.length ? 'pb-1 px-2 pt-2' : ''
			)}
			aria-labelledby={label}
			data-testid="tw-badge-input-list"
			ref={listRef}
		>
			{badges.map((badge) => (
				<Badge key={badge.id} badge={badge} listRef={listRef} />
			))}
		</ul>
	);
};

export interface BadgeProps {
	badge: BadgeItem;
	listRef: RefObject<HTMLUListElement>;
}

export const Badge = ({ badge, listRef }: BadgeProps) => {
	return (
		<EditPopover badge={badge} listRef={listRef}>
			<li className="relative inline-flex items-center py-0.5 px-1 rounded bg-badge-0 text-[0.875rem] font-medium leading-[1.125rem]">
				{badge.value}
			</li>
		</EditPopover>
	);
};
