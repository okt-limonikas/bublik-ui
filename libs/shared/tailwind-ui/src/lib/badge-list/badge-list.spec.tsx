/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { it, describe, expect } from 'vitest';
import { render, fireEvent } from '@testing-library/react';

import { BadgeList, BadgeListProps } from './badge-list';

const getBadgeListProps = (): BadgeListProps => {
	return {
		badges: [
			{ payload: 'badge-1' },
			{ isImportant: true, payload: 'badge-important' },
			{ payload: 'badge-3' }
		]
	};
};

describe('components/BadgeList', () => {
	it('should render successfully', () => {
		const { getByTestId } = render(<BadgeList {...getBadgeListProps()} />);

		const badge = getByTestId('tw-badge-list');

		expect(badge).toBeVisible();
	});

	it('should match snapshot', () => {
		const { asFragment } = render(<BadgeList {...getBadgeListProps()} />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should accept background color for badges', () => {
		const { asFragment } = render(
			<BadgeList {...getBadgeListProps()} className="bg-badge-10" />
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should highlight selected badges', () => {
		const { asFragment } = render(
			<BadgeList {...getBadgeListProps()} selectedBadges={['badge-3']} />
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should highlight selected badges with custom background color', () => {
		const { asFragment } = render(
			<BadgeList
				{...getBadgeListProps()}
				selectedBadges={['badge-3']}
				className="bg-badge-10"
			/>
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should render env badges', () => {
		const { asFragment } = render(
			<BadgeList
				{...getBadgeListProps()}
				badges={[{ payload: 'badge-1' }, { payload: 'env=test-env' }]}
				selectedBadges={['badge-3']}
				className="bg-badge-10"
			/>
		);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should call onClick with badge item payload', async () => {
		const mock = vi.fn();

		const { getByText } = render(
			<BadgeList {...getBadgeListProps()} onBadgeClick={mock} />
		);

		const badge = getByText('badge-1');
		fireEvent.click(badge);
		expect(mock).toBeCalledWith({ payload: 'badge-1' });
	});
});
