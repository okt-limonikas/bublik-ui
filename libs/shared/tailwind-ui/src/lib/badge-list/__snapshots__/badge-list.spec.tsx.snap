// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/BadgeList > should accept background color for badges 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap gap-1"
    data-testid="tw-badge-list"
  >
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-10"
      data-testid="tw-badge"
    >
      badge-1
    </div>
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-6"
      data-testid="tw-badge"
    >
      badge-important
    </div>
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-10"
      data-testid="tw-badge"
    >
      badge-3
    </div>
  </div>
</DocumentFragment>
`;

exports[`components/BadgeList > should highlight selected badges 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap gap-1"
    data-testid="tw-badge-list"
  >
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-0"
      data-testid="tw-badge"
    >
      badge-1
    </div>
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-6"
      data-testid="tw-badge"
    >
      badge-important
    </div>
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-primary-wash border-primary"
      data-testid="tw-badge"
    >
      badge-3
    </div>
  </div>
</DocumentFragment>
`;

exports[`components/BadgeList > should highlight selected badges with custom background color 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap gap-1"
    data-testid="tw-badge-list"
  >
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-10"
      data-testid="tw-badge"
    >
      badge-1
    </div>
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-6"
      data-testid="tw-badge"
    >
      badge-important
    </div>
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-primary-wash border-primary"
      data-testid="tw-badge"
    >
      badge-3
    </div>
  </div>
</DocumentFragment>
`;

exports[`components/BadgeList > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap gap-1"
    data-testid="tw-badge-list"
  >
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-0"
      data-testid="tw-badge"
    >
      badge-1
    </div>
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-6"
      data-testid="tw-badge"
    >
      badge-important
    </div>
    <div
      class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-0"
      data-testid="tw-badge"
    >
      badge-3
    </div>
  </div>
</DocumentFragment>
`;

exports[`components/BadgeList > should render env badges 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-1"
    data-testid="tw-badge-list"
  >
    <div
      class="flex flex-wrap gap-1"
    >
      <div
        class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-1"
        data-testid="tw-badge"
      >
        env=test-env
      </div>
    </div>
    <div
      class="flex flex-wrap gap-1"
    >
      <div
        class="inline-flex items-center w-fit py-0.5 px-2 rounded border border-transparent text-[0.75rem] font-medium transition-colors overflow-wrap-anywhere bg-badge-10"
        data-testid="tw-badge"
      >
        badge-1
      </div>
    </div>
  </div>
</DocumentFragment>
`;
